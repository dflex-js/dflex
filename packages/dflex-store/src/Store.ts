import Generator from "@dflex/dom-gen";

import { DFlexNode } from "@dflex/core-instance";
import type { IDFlexNode, INodeInput } from "@dflex/core-instance";

import type { RegisterInputSuper } from "./types";

class Store {
  registry: {
    [id: string]: IDFlexNode;
  };

  DOMGen: Generator;

  private _lastKeyIdentifier: string | null;

  constructor() {
    this._lastKeyIdentifier = null;

    this.registry = {};

    this.DOMGen = new Generator();
  }

  private _submitElementToRegistry(element: RegisterInputSuper) {
    const { id, depth, isPaused, isInitialized, readonly, ...rest } = element;

    const { order, keys } = this.DOMGen.register(id, depth);

    const coreElement: INodeInput = {
      id,
      order,
      keys,
      depth,
      readonly: !!readonly,
      ...rest,
    };

    this.registry[id] = new DFlexNode(coreElement, {
      isInitialized,
      isPaused,
    });
  }

  register(element: RegisterInputSuper) {
    /**
     * Using parentID, because it's impossible to know if this element belongs
     * to the same branch or not.
     */
    if (element.parentID) {
      // This is the first element in the branch then.
      if (!this._lastKeyIdentifier) {
        this._lastKeyIdentifier = element.parentID;
        // Change means new branch.
      } else if (element.parentID !== this._lastKeyIdentifier) {
        const { id, depth, ...rest } = element;
        // Create a fake parent node to close the branch.
        this._submitElementToRegistry({
          id: element.parentID,
          depth: depth + 1,
          ...rest,
        });
      }
    }

    this._submitElementToRegistry(element);
  }

  /**
   * Gets all element IDs Siblings in given node represented by sibling key.
   *
   * @param siblingsKy -
   */
  getElmBranchByKey(siblingsKy: string) {
    return this.DOMGen.getElmBranch(siblingsKy);
  }

  unregister(id: string) {
    delete this.registry[id];
  }

  destroyBranch(branchKey: string) {
    this.DOMGen.destroyBranch(branchKey, (id) => {
      this.unregister(id);
    });
  }

  destroy() {
    Object.keys(this.DOMGen.branches).forEach((branchKey) => {
      this.DOMGen.destroyBranch(branchKey, (id) => {
        delete this.registry[id];
      });
    });

    this.DOMGen.destroy();
  }
}

export default Store;
