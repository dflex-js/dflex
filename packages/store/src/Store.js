import Generator from "@dflex/dom-gen/src";
import AbstractStore from "./AbstractStore";
import Tracker from "./Tracker";

/**
 * Store class contains all dnd elements and their orders.
 *
 * @extends {Generator}
 *
 */
class Store {
  constructor() {
    this.abstractStore = new AbstractStore();
    this.DOMGen = new Generator();

    this.tracker = new Tracker();
  }

  /**
   * Add elements to registry.
   *
   * @param {Object} elmInstance
   * @param {function} CustomInstance - constructor function.
   * @param {Object} opts - extra options to be stored in the registry.
   * @memberof Store
   */
  register(elmInstance, CustomInstance, opts) {
    const { id, depth, ...restOfElmInstance } = elmInstance;

    const elementRecord = this.abstractStore.registry[id];

    let coreInstance;

    if (elementRecord) {
      /**
       * Element already exist with all properties
       */
      if (elementRecord.element) return;

      /**
       * Element has been in and rested. Create new coreInstance
       */
      coreInstance = Object.assign(restOfElmInstance, elementRecord, opts);
    } else {
      /**
       * Element is completely new.
       */
      const pointer = this.DOMGen.getElmPointer(id, depth);

      coreInstance = Object.assign(elmInstance, pointer, opts);
    }

    // Register the element when we have coreInstance
    if (coreInstance) this.abstractStore.register(coreInstance, CustomInstance);
  }

  /**
   * Cleans up element reference and any associated properties.
   * Preserve element pointer.
   *
   * @param {string} elmId
   * @memberof Store
   */
  cleanup(elmId) {
    this.abstractStore.resetElmById(elmId);
  }

  /**
   * Gets element from registry by Id.
   *
   * @param {string} id
   * @returns {elmInstance} - elmInstance Object.
   * @memberof Store
   */
  getElmById(id) {
    return this.abstractStore.getElmById(id);
  }

  /**
   * Gets all element IDs Siblings in given node represented by sibling key.
   *
   * @param {string} key
   * @returns {string|Array} - elements siblings list.
   * @memberof Store
   */
  getElmBranchByKey(ky) {
    return this.DOMGen.getElmBranch(ky);
  }

  /**
   * Gets element connections instance for a given id.
   *
   * @param {string} elmId
   *
   * @returns {Object}  connectObj
   * @returns {elmInstance} connectObj.element  - Targeted element.
   * @returns {elmInstance} connectObj.parent - Container element.
   * @returns {Object} connectObj.branches - Container element.
   * @returns {Array} branches.siblings - Array contains ids, siblings to element-id.
   * @returns {Array} branches.parents - Array contains ids, parents to element-id.
   *
   * @memberof Store
   */
  getElmTreeById(elmId) {
    const element = this.getElmById(elmId);

    const {
      keys: { sK, pK },
      order: { parent: pi },
    } = element;

    /**
     * getting connected branches
     */
    const siblings = this.getElmBranchByKey(sK);
    const parents = this.getElmBranchByKey(pK);

    /**
     * getting parent instance
     */
    let parent = null;
    if (parents !== undefined) {
      const parentsID = Array.isArray(parents) ? parents[pi] : parents;
      parent = this.getElmById(parentsID);
    }

    return {
      element,
      parent,

      branches: {
        siblings,
        parents,
      },
    };
  }
}

export default Store;
