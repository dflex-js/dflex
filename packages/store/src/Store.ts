import Generator, { ELmBranch } from "@dflex/dom-gen";

import { Class, ElmInstance, ElmWIthPointer } from "./interfaces";

class Store<T> {
  registry: {
    [id: string]: T;
  };

  DOMGen: Generator;

  constructor() {
    this.registry = {};

    this.DOMGen = new Generator();
  }

  /**
   * Delete element from the registry. Should be called only when element is
   * unmounted and expected to return with different positions only. Otherwise,
   * call `detachElmRef.`
   *
   * @param id -
   */
  deleteElm(id: string) {
    const { [id]: oldRecord, ...rest } = this.registry;

    this.registry = rest;
  }

  /**
   * Mutate elmInstance into CustomInstance then add the new object to registry
   * by id.
   *
   * @param element -
   * @param CustomInstance -
   */
  register(element: ElmInstance, CustomInstance?: Class<T>) {
    const { id, depth, ref } = element;

    if (!ref) return;

    const { order, keys } = this.DOMGen.getElmPointer(id, depth);

    const coreElement: ElmWIthPointer = { id, depth, ref, order, keys };

    // @ts-ignore
    this.registry[id] =
      CustomInstance && typeof CustomInstance.constructor === "function"
        ? new CustomInstance(coreElement)
        : coreElement;
  }

  /**
   * Gets element from registry by Id.
   *
   * @param id -
   */
  getElmById(id: string) {
    return this.registry[id];
  }

  /**
   * Gets all element IDs Siblings in given node represented by sibling key.
   *
   * @param ky -
   */
  getElmBranchByKey(ky: string): ELmBranch {
    return this.DOMGen.getElmBranch(ky);
  }
}

export default Store;
