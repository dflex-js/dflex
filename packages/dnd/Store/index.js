import Store from "@dflex/store/src";
import CoreInstance from "@dflex/core-instance/src";

class DnDStoreImp extends Store {
  /**
   * Register element for Store
   *
   * @param {Object} elmInstance
   * @param {number} elmInstance.depth
   * @param {string} elmInstance.id
   * @param {node} elmInstance.element
   * @memberof DraggableStoreImp
   */
  register(elmInstance) {
    super.register(elmInstance, CoreInstance);
  }
}

let store;

// eslint-disable-next-line func-names
export default (function () {
  if (!store) {
    store = new DnDStoreImp();
  }

  return store;
})();
