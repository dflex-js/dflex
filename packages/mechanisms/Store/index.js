import Store from "@dflex/store/src";
import CoreInstance from "@dflex/core-instance/src";

class DnDStoreImp extends Store {
  register(elmInstance) {
    super.register(elmInstance, CoreInstance);
  }
}

let store;

// eslint-disable-next-line func-names
(function () {
  if (!store) {
    store = new DnDStoreImp();
  }

  return store;
})();
