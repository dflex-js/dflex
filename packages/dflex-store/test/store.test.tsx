/* eslint-disable import/no-extraneous-dependencies */
import * as React from "react";
import * as ReactTestUtils from "react-dom/test-utils";
import { Root, createRoot } from "react-dom/client";

import Store from "../src";

// TODO: Add more testing cases to cover all store methods.

describe("Testing Store Package", () => {
  const store = new Store();

  const elm0D0 = {
    id: "id-0",
    depth: 0,
    parentID: "p-id-0",
    readonly: false,
    isInitialized: true,
    isPaused: false,
    scrollX: 0,
    scrollY: 0,
  };

  const elm1D0 = {
    id: "id-1",
    depth: 0,
    parentID: "p-id-0",
    readonly: false,
    isInitialized: true,
    isPaused: false,
    scrollX: 0,
    scrollY: 0,
  };

  const elm2D0 = {
    id: "id-2",
    depth: 0,
    parentID: "p-id-0",
    readonly: false,
    isInitialized: true,
    isPaused: false,
    scrollX: 0,
    scrollY: 0,
  };

  const elm0D1 = {
    id: "p-id-0",
    depth: 1,
    parentID: "",
    readonly: false,
    isInitialized: true,
    isPaused: false,
    scrollX: 0,
    scrollY: 0,
  };

  // const elmThrows = {
  //   id: "id-not-attached",
  //   depth: 1,
  //   parentID: "",
  //   readonly: false,
  //   isInitialized: true,
  //   isPaused: false,
  //   scrollX: 0,
  //   scrollY: 0,
  // };

  let container: HTMLDivElement | null;
  let reactRoot: Root;

  beforeAll(() => {
    container = document.createElement("div");
    reactRoot = createRoot(container);
    document.body.appendChild(container);

    function init() {
      const ref = React.createRef<HTMLDivElement>();

      function TestBase() {
        React.useEffect(() => {
          if (ref.current) {
            [elm0D0, elm1D0, elm2D0, elm0D1].forEach((elm) => {
              store.register(elm);
            });
          }

          return () => {
            [elm0D0, elm1D0, elm2D0, elm0D1].forEach((elm) => {
              store.unregister(elm.id);
            });
          };
        }, [ref]);

        return (
          <div ref={ref} id={elm0D1.id}>
            {[elm0D0, elm1D0, elm2D0].map((elm) => (
              <div id={elm.id} key={elm.id} />
            ))}
          </div>
        );
      }

      ReactTestUtils.act(() => {
        reactRoot.render(<TestBase />);
      });
    }

    init();
  });

  afterAll(() => {
    document.body.removeChild(container!);
    container = null;
  });

  it("Registers new elements", () => {
    // @ts-ignore - Just for testing purposes.
    expect(store.DOMGen.getElmBranchByKey("0-0")).toStrictEqual([
      "id-0",
      "id-1",
      "id-2",
    ]);

    // @ts-ignore - Just for testing purposes.
    expect(store.DOMGen.getElmBranchByKey("1-1")).toStrictEqual(["p-id-0"]);

    // @ts-ignore - Just for testing purposes.
    expect(store.DOMGen.getElmBranchByKey("2-0")).toStrictEqual([""]);
  });

  it("Snaps shot registry", () => {
    expect(store.registry).toMatchSnapshot();
  });

  it("Returns element instance by element id", () => {
    const elemInstance = store.registry.get(elm0D0.id);

    expect(elemInstance).toMatchSnapshot();
  });

  it("Returns element branch", () => {
    const elemInstance = store.registry.get(elm0D0.id)!;

    const elemBranch = store.getElmBranchByKey(elemInstance.keys.SK);

    expect(elemBranch).toStrictEqual(["id-0", "id-1", "id-2"]);
  });

  it("Snaps shot registry after delete", () => {
    expect(store.registry).toMatchSnapshot();
  });

  // it("Throws because it can't find the reference", () => {
  //   expect(() => store.register(elmThrows)).toThrow();
  // });
});
