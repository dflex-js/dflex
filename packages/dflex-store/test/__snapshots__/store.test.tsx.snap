// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Testing Store Package Returns element instance by element id 1`] = `
DFlexNode {
  "animatedFrame": null,
  "depth": 0,
  "id": "id-0",
  "isInitialized": false,
  "isPaused": true,
  "isVisible": false,
  "keys": Object {
    "CHK": null,
    "PK": "1-0",
    "SK": "0-0",
  },
  "order": Object {
    "parent": 0,
    "self": 0,
  },
  "readonly": false,
  "ref": null,
}
`;

exports[`Testing Store Package Snaps shot registry 1`] = `
Object {
  "id-0": DFlexNode {
    "animatedFrame": null,
    "depth": 0,
    "id": "id-0",
    "isInitialized": false,
    "isPaused": true,
    "isVisible": false,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "order": Object {
      "parent": 0,
      "self": 0,
    },
    "readonly": false,
    "ref": null,
  },
  "id-1": DFlexNode {
    "animatedFrame": null,
    "depth": 0,
    "id": "id-1",
    "isInitialized": false,
    "isPaused": true,
    "isVisible": false,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "order": Object {
      "parent": 0,
      "self": 1,
    },
    "readonly": false,
    "ref": null,
  },
  "id-2": DFlexNode {
    "animatedFrame": null,
    "depth": 0,
    "id": "id-2",
    "isInitialized": false,
    "isPaused": true,
    "isVisible": false,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "order": Object {
      "parent": 0,
      "self": 2,
    },
    "readonly": false,
    "ref": null,
  },
  "p-id-0": DFlexNode {
    "animatedFrame": null,
    "depth": 1,
    "id": "p-id-0",
    "isInitialized": false,
    "isPaused": true,
    "isVisible": false,
    "keys": Object {
      "CHK": "2-0",
      "PK": "2-0",
      "SK": "1-1",
    },
    "order": Object {
      "parent": 1,
      "self": 0,
    },
    "readonly": false,
    "ref": null,
  },
}
`;

exports[`Testing Store Package Snaps shot registry after delete 1`] = `
Object {
  "id-0": DFlexNode {
    "animatedFrame": null,
    "depth": 0,
    "id": "id-0",
    "isInitialized": false,
    "isPaused": true,
    "isVisible": false,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "order": Object {
      "parent": 0,
      "self": 0,
    },
    "readonly": false,
    "ref": null,
  },
  "id-1": DFlexNode {
    "animatedFrame": null,
    "depth": 0,
    "id": "id-1",
    "isInitialized": false,
    "isPaused": true,
    "isVisible": false,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "order": Object {
      "parent": 0,
      "self": 1,
    },
    "readonly": false,
    "ref": null,
  },
  "id-2": DFlexNode {
    "animatedFrame": null,
    "depth": 0,
    "id": "id-2",
    "isInitialized": false,
    "isPaused": true,
    "isVisible": false,
    "keys": Object {
      "CHK": null,
      "PK": "1-0",
      "SK": "0-0",
    },
    "order": Object {
      "parent": 0,
      "self": 2,
    },
    "readonly": false,
    "ref": null,
  },
  "p-id-0": DFlexNode {
    "animatedFrame": null,
    "depth": 1,
    "id": "p-id-0",
    "isInitialized": false,
    "isPaused": true,
    "isVisible": false,
    "keys": Object {
      "CHK": "2-0",
      "PK": "2-0",
      "SK": "1-1",
    },
    "order": Object {
      "parent": 1,
      "self": 0,
    },
    "readonly": false,
    "ref": null,
  },
}
`;
